mw-chat/internal/services/group_rooms.service.go

```diff
@@ -1,56 +1,168 @@
+	groupRoomsRepository GroupRoomsRepository
 }
 
-func (cc *GroupRoomsService) GetGroupRoomsPreview() {
-	// ctx.JSON(http.StatusOK, &schemas.GetRoomsResponse{})
+func NewGroupRoomsService(pool *pgxpool.Pool, groupRoomsRepository GroupRoomsRepository) *GroupRoomsService {
+	return &GroupRoomsService{pool, groupRoomsRepository}
 }
 
-func (cc *GroupRoomsService) GetGroupRoomById() {
-	// groupRoomId := ctx.Param("groupRoomId")
-	// fmt.Println(groupRoomId)
-
-	// ctx.JSON(http.StatusOK, &schemas.RoomPopulatedResponse{})
+func (groupRoomsService *GroupRoomsService) GetGroupRoomsPreview(ctx context.Context, userUuid uuid.UUID) (*schemas.GetRoomsResponse, error) {
+	groupRoomsRaw, err := groupRoomsService.groupRoomsRepository.GetGroupRoomsByUserUUID(ctx, pgtype.UUID{Bytes: userUuid})
+	if err != nil {
+		return nil, err
+	}
+
+	rooms := lo.Map(groupRoomsRaw, func(dbRoom db.GetGroupRoomsByUserUUIDRow, i int) schemas.RoomPreviewResponse {
+		return schemas.RoomPreviewResponse{
+			RoomID:    utils.ConvertPgUUIDToUUID(dbRoom.Uuid).String(),
+			UserID:    utils.ConvertPgUUIDToUUID(dbRoom.FirstUserUuid).String(),
+			Name:      dbRoom.Name,
+			IsBlocked: dbRoom.IsBlocked,
+		}
+	})
+
+	return &schemas.GetRoomsResponse{
+		Size:  len(rooms),
+		Rooms: rooms,
+	}, nil
 }
 
-func (cc *GroupRoomsService) CreateGroupRoom() {
-
-	// ctx.JSON(http.StatusOK, &schemas.RoomPopulatedResponse{})
+func (groupRoomsService *GroupRoomsService) GetPopulatedGroupRoom(ctx context.Context, userUUID, roomUUID uuid.UUID) (*schemas.RoomPopulatedResponse, error) {
+	params := db.GetGroupRoomByUUIDParams{
+		UserUuid: pgtype.UUID{Bytes: userUUID, Valid: true},
+		RoomUuid: pgtype.UUID{Bytes: roomUUID, Valid: true},
+	}
+
+	groupRoom, err := groupRoomsService.groupRoomsRepository.GetGroupRoomByUUID(ctx, params)
+	if err != nil {
+		return nil, err
+	}
+
+	messagesRaw, err := groupRoomsService.groupRoomsRepository.GetGroupMessagesByRoomUUID(ctx, pgtype.UUID{Bytes: roomUUID, Valid: true})
+	if err != nil {
+		return nil, err
+	}
+
+	messages := lo.Map(messagesRaw, func(dbMessage db.GetGroupMessagesByRoomUUIDRow, i int) schemas.MessageResponse {
+		return schemas.MessageResponse{
+			OwnerID: utils.ConvertPgUUIDToUUID(dbMessage.OwnerUuid).String(),
+			Message: dbMessage.Text,
+		}
+	})
+
+	return &schemas.RoomPopulatedResponse{
+		RoomID:    utils.ConvertPgUUIDToUUID(groupRoom.Uuid).String(),
+		UserID:    utils.ConvertPgUUIDToUUID(groupRoom.FirstUserUuid).String(),
+		Name:      groupRoom.Name,
+		Messages:  messages,
+		IsBlocked: groupRoom.IsBlocked,
+	}, nil
 }
 
-func (cc *GroupRoomsService) UpdateGroupRoom() {
-	// groupRoomId := ctx.Param("groupRoomId")
-	// fmt.Println(groupRoomId)
+func (groupRoomsService *GroupRoomsService) CreateGroupRoom(ctx context.Context, roomName string, userUUID uuid.UUID) (*schemas.RoomPopulatedResponse, error) {
+	now := time.Now()
+
+	tx, err := groupRoomsService.pool.Begin(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer tx.Rollback(ctx)
```

Comment:
will it be executed on all scenarios?

Reply:
Yes, I found it in sql doc. It standard scenario. When we do it after commit, it's alright.